extensions/wikia/RelatedPages/RelatedPages.class.php

```diff
@@ -232,62 +232,66 @@ protected function getPagesForCategory( $category ) {
 	 * @throws DBUnexpectedError|MWException
 	 */
 	protected function getPagesForCategories( $articleId, $limit, Array $categories ) {
-		global $wgMemc;
-
 		if ( empty( $categories ) ) {
 			return [];
 		}
 
 		wfProfileIn( __METHOD__ );
+		$fname = __METHOD__;
 
-		$cacheKey = wfMemcKey( __METHOD__, $articleId );
-		$cache = $wgMemc->get( $cacheKey );
+		// make the caching key more deterministic
+		sort( $categories );
 
-		if ( is_array( $cache ) ) {
-			wfProfileOut( __METHOD__ );
-			return $cache;
-		}
-
-		$dbr = wfGetDB( DB_SLAVE );
-		$pages = array();
+		$pages = WikiaDataAccess::cache(
+			wfMemcKey( __METHOD__, 'categories', md5( serialize( $category ) ) ),
+			WikiaResponse::CACHE_STANDARD,
+			function() use ( $categories, $fname, $limit ) {
+				$dbr = wfGetDB( DB_SLAVE );
+				$pages = [];
+
+				$tables = [ "categorylinks" ];
+				$joinSql = $this->getPageJoinSql( $dbr, $tables );
+
+				$innerSQL = $dbr->selectSQLText(
+					$tables,
+					array( "cl_from AS page_id" ),
+					array( "cl_to IN ( " . $dbr->makeList( $categories ) . " )" ),
+					$fname,
+					[
+						'LIMIT' => self::CATEGORY_LINKS_LIMIT
+					],
+					$joinSql
+				);
 
-		$tables = array( "categorylinks" );
-		$joinSql = $this->getPageJoinSql( $dbr, $tables );
+				# sanitize query parameters
+				$articleId = intval( $articleId );
+				$limit = intval( $limit );
+
+				$sql = "SELECT page_id, count(*) c FROM ( $innerSQL ) i GROUP BY page_id ORDER BY c desc LIMIT $limit";
+				$res = $dbr->query( $sql, $fname );
+				while ( $row = $dbr->fetchObject( $res ) ) {
+					$pageId = (int) $row->page_id;
+					$title = Title::newFromId( $pageId );
+
+					// filter out redirect pages (RT #72662)
+					if ( $title instanceof Title && $title->exists() && !$title->isRedirect() ) {
```

Comment:
IMHO the filter for redirect could be done on the SQL level, JOIN with `page` would be needed and condition on `page_is_redirect`, but this would save us from creating Title objects that do one query per instantiation. What do you think?